src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -0,0 +1,271 @@
+package com.mysql.cj.jdbc.ha.plugins;
+
+import com.mysql.cj.Messages;
+import com.mysql.cj.conf.*;
+import com.mysql.cj.exceptions.MysqlErrorNumbers;
+import com.mysql.cj.jdbc.JdbcConnection;
+import com.mysql.cj.log.Log;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+public class ReadWriteSplittingPlugin implements IConnectionPlugin {
+  public static final int WRITER_INDEX = 0;
+  public static final String METHOD_SET_READ_ONLY = "setReadOnly";
+
+  private final ICurrentConnectionProvider currentConnectionProvider;
+  private final IConnectionProvider connectionProvider;
+  private final PropertySet propertySet;
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+
+  private boolean inTransaction = false;
+  private boolean readOnly = false;
+  private List<HostInfo> hosts = new ArrayList<>();
+  private JdbcConnection writerConnection;
+  private JdbcConnection readerConnection;
+
+  public ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this(currentConnectionProvider, new BasicConnectionProvider(), propertySet, nextPlugin, logger);
+  }
+
+  ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      IConnectionProvider connectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this.currentConnectionProvider = currentConnectionProvider;
+    this.connectionProvider = connectionProvider;
+    this.propertySet = propertySet;
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+  }
+
+  @Override
+  public Object execute(Class<?> methodInvokeOn, String methodName, Callable<?> executeSqlFunc, Object[] args) throws Exception {
+    if(METHOD_SET_READ_ONLY.equals(methodName)) {
+      switchConnectionIfRequired((boolean) args[0]);
+    }
+    return this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc, args);
+  }
+
+  void switchConnectionIfRequired(boolean readOnly) throws SQLException {
```

Comment:
Should we take into account this.inTransaction? Switching connection while been in transaction shouldn't be allowed.

Reply:
Good point, thanks